src/ripple/overlay/impl/OverlayImpl.cpp

```diff
@@ -570,10 +609,62 @@ OverlayImpl::onPeerDeactivate (Peer::id_t id,
     m_publicKeyMap.erase(publicKey);
 }
 
-/** The number of active peers on the network
-    Active peers are only those peers that have completed the handshake
-    and are running the Ripple protocol.
-*/
+void
+OverlayImpl::onManifests (Job&,
+    std::shared_ptr<protocol::TMManifests> const& inbox,
+        std::shared_ptr<PeerImp> const& from)
+{
+    auto const n = inbox->list_size();
+    auto const& journal = from->pjournal();
+
+    if (journal.debug) journal.debug
+        << "TMManifest, " << n << (n == 1 ? " item" : " items");
+
+    protocol::TMManifests outbox;
+
+    for (std::size_t i = 0; i < n; ++i)
+    {
+        auto& s = inbox->list().Get(i).stobject();
+        STObject st(sfGeneric);
+        try
+        {
+            SerialIter sit(s.data(), s.size());
+            st.set(sit);
+        }
+        catch(...)
+        {
+            if (journal.info) journal.info
+                << "Bad manifest #" << i + 1;
+            continue;
+        }
+        auto const pk = get<AnyPublicKey>(st, sfPublicKey);
+        auto const seq = get(st, sfSequence);
+
+        if (pk && seq && manifestCache_.applyManifest (*pk, *seq, s, journal))
+        {
+            outbox.add_list()->set_stobject(s);
+        }
+    }
+
+    if (outbox.list_size() == 0)
+        return;
+
+    if (journal.debug) journal.debug
+        << "Forwarding " << outbox.list_size() << " manifests...";
+
+    auto const msg = std::make_shared<Message>(outbox, protocol::mtMANIFESTS);
+
+    for_each( [&](std::shared_ptr<PeerImp> const& peer)
+        {
+            if (peer != from)
+            {
+                if (auto const& j = peer->pjournal().debug) j
+                    << "... to " << peer.get();
+                peer->send(msg);
+            }
+        });
+}
```

Comment:
I modified `applyManifest()` so it releases the lock before verifying the signature, then re-acquires it and checks the prerequisite conditions again.